http/src/main/java/io/micronaut/http/bind/binders/BodyRequestArgumentBinder.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.bind.binders;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.bind.annotation.AbstractArgumentBinder;
+import io.micronaut.core.bind.annotation.Bindable;
+import io.micronaut.core.convert.ArgumentConversionContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.convert.value.ConvertibleValues;
+import io.micronaut.core.reflect.ClassUtils;
+import io.micronaut.core.type.Argument;
+import io.micronaut.http.HttpMethod;
+import io.micronaut.http.HttpRequest;
+
+import java.util.Optional;
+
+/**
+ * Bind the request argument value from the body.
+ *
+ * @param <T> A type
+ * @author Denis Stepanov
+ * @since 4.0.0
+ */
+@Internal
+public final class BodyRequestArgumentBinder<T>  extends AbstractArgumentBinder<T> implements RequestArgumentBinder<T>, PostponedRequestArgumentBinder {
+
+    /**
+     * Constructor.
+     *
+     * @param conversionService conversionService
+     */
+    public BodyRequestArgumentBinder(ConversionService conversionService) {
+        super(conversionService);
+    }
+
+    @Override
+    public BindingResult<T> bind(ArgumentConversionContext<T> context, HttpRequest<?> source) {
+        HttpMethod httpMethod = source.getMethod();
+        boolean permitsRequestBody = HttpMethod.permitsRequestBody(httpMethod);
+        if (!permitsRequestBody) {
+            return BindingResult.unsatisfied();
+        }
+
+        Argument<T> argument = context.getArgument();
+        Argument<T> argumentType;
+        if (argument.getType() == Optional.class) {
+            argumentType = (Argument<T>) argument.getFirstTypeVariable().orElse(argument);
+        } else {
+            argumentType = argument;
+        }
+
+        String inputName = argument.getAnnotationMetadata().stringValue(Bindable.NAME).orElse(argument.getName());
```

Comment:
Perhaps we can add a new interface:

```java
interace IntializableArgumentBindiner<T> extends ArgumentBinder<T> {
    void initialize(Argument<?> argument);
}
```

And then this logic can be moved to the part that sets up the argument binder instead of executing each time we bind.

Reply:
The binders are singletons shared with different routes, we might want to have something like `IntializableArgumentBindiner specific(..)`

Reply:
I will add this in the next PR